File: P3D-Legacy Core/Screens/Dialogues/_ChooseBox.cs
using System;

using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Graphics;

using P3D.Legacy.Core.Data;
using P3D.Legacy.Core.Entities;
using P3D.Legacy.Core.Input;
using P3D.Legacy.Core.Resources;
using P3D.Legacy.Core.Resources.Sound;
using P3D.Legacy.Core.Screens;

namespace P3D.Legacy.Core.Dialogues
{
    /*
    public class ChooseBox
    {
        public int Index { get; set; } = 0;
        public string[] Options { get; set; }


        float _positionY = 0;
        public bool ReadyForResult { get; set; }
        public int Result { get; set; }
        public int ResultId = 0;

        public bool ActionScript = false;
        public static int CancelIndex = -1;

        public FontContainer TextFont { get; set; }
        public bool Showing { get; set; }


        public bool DoDelegate = false;

        Action<Int32> _subs;

        public Entity[] UpdateEntities;

        public void Show(string[] options, Action<Int32> doSubs)
        {
            ResultId = 0;
            Options = options;
            Index = 0;
            ReadyForResult = false;
            Showing = true;
            _subs = doSubs;
            ActionScript = false;
            DoDelegate = true;
            TextFont = FontManager.GetFontContainer("textfont");

            SetupOptions();
        }
        public void Show(string[] options, int id, bool actionScript)
        {
            ResultId = id;
            Options = options;
            Index = 0;
            ReadyForResult = false;
            Showing = true;
            ActionScript = actionScript;
            DoDelegate = false;
            TextFont = FontManager.GetFontContainer("textfont");

            SetupOptions();
        }
        public void Show(string[] options, int id, Entity[] updateEntities)
        {
            ResultId = id;
            Options = options;
            Index = 0;
            ReadyForResult = false;
            Showing = true;
            UpdateEntities = updateEntities;
            ActionScript = false;
            DoDelegate = false;
            TextFont = FontManager.GetFontContainer("textfont");

            SetupOptions();
        }

        private void SetupOptions()
        {
            for (var i = 0; i <= Options.Length - 1; i++)
                Options[i] = Options[i].Replace("<playername>", Core.Player.Name);
        }

        public int GetResult(int id) => ReadyForResult ? (ResultId == id ? Result : -1) : -1;

        public void Update() => Update(true);

        public void Update(bool raiseClickEvent)
        {
            if (Showing)
            {
                if (Controls.Down(true, true, true))
                {
                    Index += 1;
                }
                if (Controls.Up(true, true, true))
                {
                    Index -= 1;
                }

                if (Index < 0)
                {
                    Index = Options.Length - 1;
                }
                if (Index == Options.Length)
                {
                    Index = 0;
                }
                if (raiseClickEvent)
                {
                    if (Controls.Accept())
                    {
                        PlayClickSound();
                        Result = Index;
                        HandleResult();
                    }
                    if (Controls.Dismiss() && CancelIndex > -1)
                    {
                        PlayClickSound();
                        Result = CancelIndex;
                        HandleResult();
                    }
                }
            }
        }

        private void PlayClickSound()
        {
            if (Screen.TextBox.Showing == false)
                SoundManager.PlaySound("select");
        }

        private void HandleResult()
        {
            CancelIndex = -1;
            ReadyForResult = true;
            Showing = false;
            if (DoDelegate)
            {
                _subs(Result);
            }
            else
            {
                if (Core.CurrentScreen.Identification == Screen.Identifications.OverworldScreen)
                {
                    if (ActionScript)
                    {
                        BaseOverworldScreen c = (BaseOverworldScreen)Core.CurrentScreen;
                        c.ActionScript.Switch(this.Options(Result));
                    }
                    else
                    {
                        foreach (Entity entity in UpdateEntities)
                        {
                            entity.ResultFunction(Result);
                        }
                    }
                }
            }
        }

        public void Draw(Vector2 position)
        {
            if (Showing)
            {
                var with1 = Core.SpriteBatch;
                with1.Draw(TextureManager.GetTexture("GUI\\Overworld\\ChooseBox"), new Rectangle(Convert.ToInt32(position.X), Convert.ToInt32(position.Y), 288, 48), new Rectangle(0, 0, 96, 16), Color.White);
                for (var i = 0; i <= Options.Length - 2; i++)
                {
                    with1.Draw(TextureManager.GetTexture("GUI\\Overworld\\ChooseBox"), new Rectangle(Convert.ToInt32(position.X), Convert.ToInt32(position.Y) + 48 + i * 48, 288, 48), new Rectangle(0, 16, 96, 16), Color.White);
                }
                with1.Draw(TextureManager.GetTexture("GUI\\Overworld\\ChooseBox"), new Rectangle(Convert.ToInt32(position.X), Convert.ToInt32(position.Y) + 96 + (Options.Length - 2) * 48, 288, 48), new Rectangle(0, 32, 96, 16), Color.White);
                for (var i = 0; i <= Options.Length - 1; i++)
                {
                    float m = 1f;
                    switch (TextFont.FontName.ToLower())
                    {
                        case "textfont":
                        case "braille":
                            m = 2f;
                            break;
                    }

                    with1.DrawString(TextFont.SpriteFont, Options[i].Replace("[POKE]", "Poké"), new Vector2(Convert.ToInt32(position.X + 40), Convert.ToInt32(position.Y) + 32 + i * 48), Color.Black, 0f, Vector2.Zero, m, SpriteEffects.None, 0f);
                }
                with1.Draw(TextureManager.GetTexture("GUI\\Overworld\\ChooseBox"), new Rectangle(Convert.ToInt32(position.X + 20), Convert.ToInt32(position.Y) + 36 + Index * 48, 10, 20), new Rectangle(96, 0, 3, 6), Color.White);
            }
        }

        public void Draw()
        {
            if (Showing)
            {
                Vector2 position = new Vector2(Convert.ToInt32(Core.WindowSize.Width / 2) - 48, Core.WindowSize.Height - 160f - 96f - (Options.Length - 1) * 48);
                Draw(position);
            }
        }

    }
    */
}

File: 2.5DHero/2.5DHero/Dialogues/ChooseBox.vb
﻿Imports P3D.Legacy.Core
Imports P3D.Legacy.Core.Data
Imports P3D.Legacy.Core.Dialogues
Imports P3D.Legacy.Core.Entities
Imports P3D.Legacy.Core.Input
Imports P3D.Legacy.Core.Resources
Imports P3D.Legacy.Core.Resources.Managers
Imports P3D.Legacy.Core.Resources.Managers.Sound
Imports P3D.Legacy.Core.Resources.Sound
Imports P3D.Legacy.Core.Screens

Public Class ChooseBox
    Implements IChooseBox

    Public Property Index As Integer Implements IChooseBox.Index
        Get
            Return _index
        End Get
        Set
            _index = Value
        End Set
    End Property

    Public Property Options As String() Implements IChooseBox.Options
        Get
            Return _options
        End Get
        Set
            _options = Value
        End Set
    End Property

    Dim PositionY As Single = 0

    Public Property ReadyForResult As Boolean Implements IChooseBox.ReadyForResult
    Public Property Result As Integer Implements IChooseBox.Result
    Public resultID As Integer = 0
    Public ActionScript As Boolean = False

    Public Shared CancelIndex As Integer = -1
    Private _options As String()
    Private _index As Integer = 0

    Public Property TextFont As FontContainer Implements IChooseBox.TextFont
    Public Property Showing As Boolean Implements IChooseBox.Showing

    Public DoDelegate As Boolean = False

    Dim Subs As Action(Of Int32)

    Public UpdateEntities() As Entity

    Public Sub Show(ByVal Options() As String, ByVal DoSubs As Action(Of Int32)) Implements IChooseBox.Show
        Me.resultID = 0
        Me.Options = Options
        Me.Index = 0
        Me.ReadyForResult = False
        Me.Showing = True
        Me.Subs = DoSubs
        Me.ActionScript = False
        Me.DoDelegate = True
        Me.TextFont = FontManager.GetFontContainer("textfont")

        SetupOptions()
    End Sub

    Public Sub Show(ByVal Options() As String, ByVal ID As Integer, ByVal ActionScript As Boolean) Implements IChooseBox.Show
        Me.resultID = ID
        Me.Options = Options
        Me.Index = 0
        Me.ReadyForResult = False
        Me.Showing = True
        Me.ActionScript = ActionScript
        Me.DoDelegate = False
        Me.TextFont = FontManager.GetFontContainer("textfont")

        SetupOptions()
    End Sub

    Public Sub Show(ByVal Options() As String, ByVal ID As Integer, ByVal UpdateEntities() As Entity) Implements IChooseBox.Show
        Me.resultID = ID
        Me.Options = Options
        Me.Index = 0
        Me.ReadyForResult = False
        Me.Showing = True
        Me.UpdateEntities = UpdateEntities
        Me.ActionScript = False
        Me.DoDelegate = False
        Me.TextFont = FontManager.GetFontContainer("textfont")

        SetupOptions()
    End Sub

    Private Sub SetupOptions()
        For i = 0 To Options.Count - 1
            Options(i) = Options(i).Replace("<playername>", Core.Player.Name)
        Next
    End Sub

    Public Function GetResult(ByVal ID As Integer) As Integer Implements IChooseBox.GetResult
        If Me.ReadyForResult = True Then
            If Me.resultID = ID Then
                Return Result
            Else
                Return -1
            End If
        Else
            Return -1
        End If
    End Function

    Public Sub Update() Implements IChooseBox.Update
        Update(True)
    End Sub

    Public Sub Update(ByVal RaiseClickEvent As Boolean) Implements IChooseBox.Update
        If Me.Showing = True Then
            If Controls.Down(True, True, True) Then
                Me.index += 1
            End If
            If Controls.Up(True, True, True) Then
                Me.index -= 1
            End If

            If Me.index < 0 Then
                Me.index = Me.Options.Count - 1
            End If
            If Me.index = Me.Options.Count Then
                Me.index = 0
            End If
            If RaiseClickEvent = True Then
                If Controls.Accept() = True Then
                    Me.PlayClickSound()
                    Me.result = index
                    Me.HandleResult()
                End If
                If Controls.Dismiss() = True And CancelIndex > -1 Then
                    Me.PlayClickSound()
                    Me.result = CancelIndex
                    Me.HandleResult()
                End If
            End If
        End If
    End Sub

    Private Sub PlayClickSound()
        If Screen.TextBox.Showing = False Then
            SoundEffectManager.PlaySound("select")
        End If
    End Sub

    Private Sub HandleResult()
        ChooseBox.CancelIndex = -1
        Me.readyForResult = True
        Me.Showing = False
        If Me.DoDelegate = True Then
            Subs(result)
        Else
            If Core.CurrentScreen.Identification = Screen.Identifications.OverworldScreen Then
                If Me.ActionScript = True Then
                    Dim c As OverworldScreen = CType(Core.CurrentScreen, OverworldScreen)
                    c.ActionScript.Switch(Me.Options(result))
                Else
                    For Each Entity As Entity In UpdateEntities
                        Entity.ResultFunction(result)
                    Next
                End If
            End If
        End If
    End Sub

    Public Sub Draw(ByVal Position As Vector2) Implements IChooseBox.Draw
        If Me.Showing = True Then
            With Core.SpriteBatch
                .Draw(TextureManager.GetTexture("GUI\Overworld\ChooseBox"), New Rectangle(CInt(Position.X), CInt(Position.Y), 288, 48), New Rectangle(0, 0, 96, 16), Color.White)
                For i = 0 To Options.Count - 2
                    .Draw(TextureManager.GetTexture("GUI\Overworld\ChooseBox"), New Rectangle(CInt(Position.X), CInt(Position.Y) + 48 + i * 48, 288, 48), New Rectangle(0, 16, 96, 16), Color.White)
                Next
                .Draw(TextureManager.GetTexture("GUI\Overworld\ChooseBox"), New Rectangle(CInt(Position.X), CInt(Position.Y) + 96 + (Options.Count - 2) * 48, 288, 48), New Rectangle(0, 32, 96, 16), Color.White)
                For i = 0 To Options.Count - 1
                    Dim m As Single = 1.0F
                    Select Case Me.TextFont.FontName.ToLower()
                        Case "textfont", "braille"
                            m = 2.0F
                    End Select

                    .DrawString(Me.TextFont.SpriteFont, Options(i).Replace("[POKE]", "Poké"), New Vector2(CInt(Position.X + 40), CInt(Position.Y) + 32 + i * 48), Color.Black, 0.0F, Vector2.Zero, m, SpriteEffects.None, 0.0F)
                Next
                .Draw(TextureManager.GetTexture("GUI\Overworld\ChooseBox"), New Rectangle(CInt(Position.X + 20), CInt(Position.Y) + 36 + index * 48, 10, 20), New Rectangle(96, 0, 3, 6), Color.White)
            End With
        End If
    End Sub

    Public Sub Draw() Implements IChooseBox.Draw
        If Me.Showing = True Then
            Dim Position As Vector2 = New Vector2(CInt(Core.windowSize.Width / 2) - 48, Core.windowSize.Height - 160.0F - 96.0F - (Options.Count - 1) * 48)
            Me.Draw(Position)
        End If
    End Sub

End Class